File: src/frontend/Lexer.cpp

#include "Lexer.h"

#include <iostream>

#include <shared/errors.h>

namespace HDLCompiler {

std::string Lexer::TokenType_to_string(Lexer::TokenType type) {
    switch (type) {
        case TokenType::None: return "(none)";
        case TokenType::EndOfFile: return "end of file";
        case TokenType::Identifier: return "identifier";
        case TokenType::Number: return "number";
        case TokenType::Dot: return "'.'";
        case TokenType::Comma: return "','";
        case TokenType::Semicolon: return "';'";
        case TokenType::Colon: return "':'";
        case TokenType::Equals: return "'='";
        case TokenType::AND: return "'&'";
        case TokenType::OR: return "'|'";
        case TokenType::XOR: return "'^'";
        case TokenType::NOT: return "'~'";
        case TokenType::LeftBrace: return "'{'";
        case TokenType::RightBrace: return "'}'";
        case TokenType::LeftBracket: return "'['";
        case TokenType::RightBracket: return "']'";
        case TokenType::LeftParenthesis: return "'('";
        case TokenType::RightParenthesis: return "')'";
        case TokenType::InKeyword: return "'in'";
        case TokenType::OutKeyword: return "'out'";
        case TokenType::BlockKeyword: return "'block'";
        default: return "INVALID";
    }
}

std::ostream& operator<<(std::ostream &stream, Lexer::TokenType type) {
    stream << Lexer::TokenType_to_string(type);
    return stream;
}

Lexer::Location::Location(unsigned long line, unsigned long column, std::string filename)
    : line(line), column(column), filename(std::move(filename)) {
    //
}

std::string Lexer::Location::to_string() const {
    return filename + ":" + std::to_string(line) + ":" + std::to_string(column);
}

std::ostream& operator<<(std::ostream &stream, const Lexer::Location &location) {
    stream << location.to_string();
    return stream;
}

Lexer::Token::Token(TokenType type, Location location)
    : type(type), location(std::move(location)), number({0, 0}) {
    //
}

std::string Lexer::Token::to_string() const {
    std::string str;
    str += Lexer::TokenType_to_string(type);
    switch (type) {
        case TokenType::Identifier:
            str += " \"" + identifier + "\"";
            break;
        case TokenType::Number:
            str += " " + std::to_string(number.value) + "#" + std::to_string(number.width);
            break;
        default:
            break;
    }
    str += " at " + location.to_string();
    return str;
}

std::ostream& operator<<(std::ostream &stream, const Lexer::Token &token) {
    stream << token.to_string();
    return stream;
}

Lexer::Lexer(std::ifstream &f, std::string filename)
    : f(f), location(1, 0, std::move(filename)) {
    //
}

int Lexer::getc() {
    lastLocation = location;

    int c = f.get();

    if (c == '\n') {
        location.line++;
        location.column = 0;
    } else {
        location.column++;
    }

    return c;
}

void Lexer::ungetc() {
    f.unget();
    location = lastLocation;
}

Lexer::Token Lexer::readNextToken() {
    int c;

    do {
        c = getc();

        // Skip comments starting with '//'
        if (c == '/') {
            if (getc() == '/') {
                // Comments end with a line break
                do {
                    c = getc();
                } while (c != '\n');
            } else {
                ungetc();
            }
        }
    } while (isspace(c));

    if (isdigit(c)) {
        std::uint64_t num = 0;

        do {
            num *= 10;
            num += c - '0';
            c = getc();
        } while (isdigit(c));

        std::uint64_t width = 0;

        if (c == '#') {
            c = getc();

            do {
                width *= 10;
                width += c - '0';
                c = getc();
            } while (isdigit(c));

            if (width == 0) {
                throw CompilerError(location.to_string() + ": Number literal width is 0");
            }
        }

        if (width > 64) {
            throw CompilerError(location.to_string() + ": Number literal width > 64");
        }

        if (width > 0 && num > (1u << width) - 1) {
            std::cerr << location << ": Number literal value doesn't fit in specified width (warning)\n";
        }

        ungetc();

        Token token(TokenType::Number, location);
        token.number.value = num;
        token.number.width = width;

        return token;
    } else if (isalpha(c) || c == '_') {
        std::string str;

        do {
            str += char(c);
            c = getc();
        } while (isalpha(c) || isdigit(c) || c == '_');

        ungetc();

        if (str == "in") {
            return Token(TokenType::InKeyword, location);
        } else if (str == "out") {
            return Token(TokenType::OutKeyword, location);
        } else if (str == "block") {
            return Token(TokenType::BlockKeyword, location);
        }
        
        Token token(TokenType::Identifier, location);
        token.identifier = str;

        return token;
    } else if (c == '.') {
        return Token(TokenType::Dot, location);
    } else if (c == ',') {
        return Token(TokenType::Comma, location);
    } else if (c == ';') {
        return Token(TokenType::Semicolon, location);
    } else if (c == ':') {
        return Token(TokenType::Colon, location);
    } else if (c == '=') {
        return Token(TokenType::Equals, location);
    } else if (c == '&') {
        return Token(TokenType::AND, location);
    } else if (c == '|') {
        return Token(TokenType::OR, location);
    } else if (c == '^') {
        return Token(TokenType::XOR, location);
    } else if (c == '~') {
        return Token(TokenType::NOT, location);
    } else if (c == '{') {
        return Token(TokenType::LeftBrace, location);
    } else if (c == '}') {
        return Token(TokenType::RightBrace, location);
    } else if (c == '[') {
        return Token(TokenType::LeftBracket, location);
    } else if (c == ']') {
        return Token(TokenType::RightBracket, location);
    } else if (c == '(') {
        return Token(TokenType::LeftParenthesis, location);
    } else if (c == ')') {
        return Token(TokenType::RightParenthesis, location);
    } else if (c == EOF) {
        return Token(TokenType::EndOfFile, location);
    }

    throw CompilerError(location.to_string() + ": Invalid character '" + (char) c + "'");
}

}

File: src/lexer.rs

use std::io::{BufReader, Read};
use std::{io, fmt};

use failure::Fail;

use crate::char_reader::CharReader;
use crate::ext_char::ExtChar;
use crate::ext_char::ExtChar::*;

#[derive(Debug, Fail)]
pub enum LexerError {
    #[fail(display = "unexpected character: '{}' at {}", _0, _1)]
    UnexpectedCharacter(char, Location),

    #[fail(display = "unexpected end of file at {}", _0)]
    UnexpectedEndOfFile(Location),

    #[fail(display = "number literal has no width specified after width separator at {}", _0)]
    NumberLiteralNoWidth(Location),

    #[fail(display = "number literal with width of 0 at {}", _0)]
    NumberLiteralWidthZero(Location),

    #[fail(display = "number literal with width greater than 64 at {}", _0)]
    NumberLiteralWidthTooBig(Location),

    #[fail(display = "number literal with value that doesn't fit into a register of specified width at {}", _0)]
    NumberLiteralValueTooBig(Location),

    #[fail(display = "{}", _0)]
    Io(#[cause] io::Error),
}

/// Trait for the Lexer to implement.
///
/// This simplifies testing the parser a lot, as it can be tested completely separate from the lexer
/// implementation
pub trait ILexer {
//    type Token = Token;
//    type Error = LexerError;

    fn get_token(&mut self) -> Result<Token, LexerError>;
}

/// Location of a token in the source code
#[derive(Copy, Clone, Debug, PartialEq)]
pub struct Location {
    pub line: usize,
    pub column: usize,
}

impl fmt::Display for Location {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        write!(f, "{}:{}", self.line, self.column)
    }
}

impl Location {
    fn new(line: usize, column: usize) -> Location {
        Location {
            line,
            column,
        }
    }
}

/// All the possible token types with associated data
#[derive(Debug, PartialEq)]
pub enum TokenData {
    EndOfFile,

    Identifier(String),
    Number { value: u64, width: Option<u64> },

    Dot,
    Comma,
    Semicolon,
    Colon,
    Equals,

    AND,
    OR,
    XOR,
    NOT,

    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    LeftParenthesis,
    RightParenthesis,

    InKeyword,
    OutKeyword,
    BlockKeyword,
}

/// Tokens of this type are emitted by the lexer
#[derive(Debug, PartialEq)]
pub struct Token {
    pub data: TokenData,
    pub location: Location,
}

impl Token {
    /// Construct a new token
    fn new(data: TokenData, location: Location) -> Token {
        Token {
            data,
            location,
        }
    }
}

/// Lexer implementation
pub struct Lexer<R: Read> {
    /// Buffered source
    reader: CharReader<BufReader<R>>,

    /// Current location in the source code
    location: Location,

    /// If this field is `Some(_)`, it is returned by `get_char()` before a new character is read
    /// from `reader`
    next_char: Option<ExtChar>,

    /// Character that was returned by the last call to `get_char()`.
    /// This is used by `unget_char()`.
    previous_char: Option<ExtChar>,
    /// Location of `previous_char`
    previous_location: Location,
}

impl<R: Read> ILexer for Lexer<R> {
//    type Token = Token;
//    type Error = LexerError;

    /// Reads from the source stream until a token is complete and returns it
    fn get_token(&mut self) -> Result<Token, LexerError> {
        let mut c;

        // Skip whitespace
        loop {
            c = self.get_char()?;

            match c {
                Char(x) if !x.is_whitespace() => break,
                EOF => break,
                _ => {}
            }
        }

        let token_location = self.location.clone();

        match c {
            // Simple tokens
            Char('.') => Ok(Token::new(TokenData::Dot, token_location)),
            Char(',') => Ok(Token::new(TokenData::Comma, token_location)),
            Char(';') => Ok(Token::new(TokenData::Semicolon, token_location)),
            Char(':') => Ok(Token::new(TokenData::Colon, token_location)),
            Char('=') => Ok(Token::new(TokenData::Equals, token_location)),
            Char('&') => Ok(Token::new(TokenData::AND, token_location)),
            Char('|') => Ok(Token::new(TokenData::OR, token_location)),
            Char('^') => Ok(Token::new(TokenData::XOR, token_location)),
            Char('~') => Ok(Token::new(TokenData::NOT, token_location)),
            Char('{') => Ok(Token::new(TokenData::LeftBrace, token_location)),
            Char('}') => Ok(Token::new(TokenData::RightBrace, token_location)),
            Char('[') => Ok(Token::new(TokenData::LeftBracket, token_location)),
            Char(']') => Ok(Token::new(TokenData::RightBracket, token_location)),
            Char('(') => Ok(Token::new(TokenData::LeftParenthesis, token_location)),
            Char(')') => Ok(Token::new(TokenData::RightParenthesis, token_location)),
            // Tokens that can't be match'ed
            Char(x) => {
                if x.is_alphabetic() || x == '_' {
                    // Identifier & keywords

                    let mut s = x.to_string();

                    loop {
                        c = self.get_char()?;

                        match c {
                            Char(x) => {
                                if x.is_alphanumeric() || x == '_' {
                                    s.push(x);
                                } else {
                                    self.unget_char();
                                    break;
                                }
                            },
                            EOF => break,
                        }
                    }

                    match s.as_ref() {
                        "in" => Ok(Token::new(TokenData::InKeyword, token_location)),
                        "out" => Ok(Token::new(TokenData::OutKeyword, token_location)),
                        "block" => Ok(Token::new(TokenData::BlockKeyword, token_location)),
                        _ => Ok(Token::new(TokenData::Identifier(s), token_location))
                    }
                } else if x.is_numeric() {
                    // Number

                    let value = self.parse_number(&mut c)?.unwrap();

                    let width = if c == Char('#') {
                        c = self.get_char()?;

                        let width = match self.parse_number(&mut c)? {
                            Some(v) => v,
                            None => return Err(LexerError::NumberLiteralNoWidth(token_location)),
                        };

                        if width == 0 {
                            return Err(LexerError::NumberLiteralWidthZero(token_location));
                        }

                        if width > 64 {
                            return Err(LexerError::NumberLiteralWidthTooBig(token_location));
                        }

                        if value > (1 << width) - 1 {
                            return Err(LexerError::NumberLiteralValueTooBig(token_location));
                        }

                        Some(width)
                    } else {
                        None
                    };

                    self.unget_char();

                    Ok(Token::new(TokenData::Number { value, width }, token_location))
                } else {
                    Err(LexerError::UnexpectedCharacter(x, token_location))
                }
            },
            // EndOfFile
            EOF => Ok(Token::new(TokenData::EndOfFile, token_location)),
        }
    }
}

impl<R: Read> Lexer<R> {
    /// Construct a new instance
    pub fn new(source: R) -> Lexer<R> {
        Lexer {
            reader: CharReader::new(BufReader::new(source)),
            location: Location::new(1, 0),
            next_char: None,
            previous_char: None,
            previous_location: Location::new(1, 0),
        }
    }

    /// Retrieve a single character from the source
    ///
    /// This method also keeps track of the current source code line and column number which is
    /// stored in `self.location`.
    /// When the end of the file is reached, this method returns `None`.
    fn get_char(&mut self) -> Result<ExtChar, LexerError> {
        let result = match self.next_char {
            Some(c) => {
                self.next_char = None;
                Ok(c)
            },
            None => self.reader.read_char(),
        };

        if let Ok(c) = result {
            self.previous_location = self.location;
            self.previous_char = Some(c);

            if let Char(c) = c {
                if c == '\n' {
                    self.location.line += 1;
                    self.location.column = 0;
                } else {
                    self.location.column += 1;
                }
            }
        }

        result.map_err(|e| LexerError::Io(e))
    }

    /// Undo the last call to `get_char()`
    ///
    /// If this method is called twice in a row (without calling `get_char()` in between), it will
    /// panic.
    fn unget_char(&mut self) {
        if self.next_char.is_some() {
            panic!("unget_char() called twice in a row");
        }

        match self.previous_char {
            Some(c) => {
                self.previous_char = None;
                self.next_char = Some(c);
                self.location = self.previous_location;
            },
            None => panic!("unget_char() called but no char to unget"),
        }
    }

    fn parse_number(&mut self, c: &mut ExtChar) -> Result<Option<u64>, LexerError> {
        if c.is_eof() || !c.get_char().unwrap().is_numeric() {
            return Ok(None);
        }

        let mut value: u64 = 0;

        loop {
            let digit = c.get_char().unwrap().to_digit(10).unwrap() as u64;

            value *= 10;
            value += digit;

            *c = self.get_char()?;

            match c {
                Char(x) if !x.is_numeric() => break,
                EOF => break,
                _ => {},
            }
        }

        Ok(Some(value))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use std::io::Cursor;

    use matches::assert_matches;

    fn expect_tokens(source_text: &str, expected_tokens: &Vec<Token>) -> Result<(), LexerError> {
        let source = Cursor::new(source_text);

        let mut lexer = Lexer::new(source);

        let mut tokens = Vec::new();

        let mut terminate = false;

        while !terminate {
            let token = lexer.get_token()?;

            if token.data == TokenData::EndOfFile {
                terminate = true;
            }

            tokens.push(token);
        }

        assert_eq!(expected_tokens, &tokens);

        Ok(())
    }

    #[test]
    fn identifier_with_digits_and_underscores() {
        let source_text = "_the_answer_is_42";
        let expected_tokens = vec![
            Token::new(TokenData::Identifier("_the_answer_is_42".to_string()), Location::new(1, 1)),
            Token::new(TokenData::EndOfFile, Location::new(1, 17)),
        ];

        let result = expect_tokens(source_text, &expected_tokens);
        assert!(result.is_ok());
    }

    #[test]
    fn number_zeroes() {
        let source_text = "00000000000000000000000005";
        let expected_tokens = vec![
            Token::new(TokenData::Number { value: 5, width: None }, Location::new(1, 1)),
            Token::new(TokenData::EndOfFile, Location::new(1, 26)),
        ];

        let result = expect_tokens(source_text, &expected_tokens);
        assert!(result.is_ok());
    }

    #[test]
    fn number_big() {
        let source_text = "9223372036854775807";
        let expected_tokens = vec![
            Token::new(TokenData::Number { value: 9223372036854775807, width: None }, Location::new(1, 1)),
            Token::new(TokenData::EndOfFile, Location::new(1, 19)),
        ];

        let result = expect_tokens(source_text, &expected_tokens);
        assert!(result.is_ok());
    }

    #[test]
    fn number_width() {
        let source_text = "42#8";
        let expected_tokens = vec![
            Token::new(TokenData::Number { value: 42, width: Some(8) }, Location::new(1, 1)),
            Token::new(TokenData::EndOfFile, Location::new(1, 4)),
        ];

        let result = expect_tokens(source_text, &expected_tokens);
        assert!(result.is_ok());
    }

    #[test]
    fn number_no_width() {
        let source_text = "42#";

        let source = Cursor::new(source_text);

        let mut lexer = Lexer::new(source);

        let result = lexer.get_token();

        assert_matches!(result, Err(LexerError::NumberLiteralNoWidth(_)));

        if let Err(LexerError::NumberLiteralNoWidth(l)) = result {
            assert_eq!(Location::new(1, 1), l);
        }
    }

    #[test]
    fn number_zero_width() {
        let source_text = "42#0";

        let source = Cursor::new(source_text);

        let mut lexer = Lexer::new(source);

        let result = lexer.get_token();

        assert_matches!(result, Err(LexerError::NumberLiteralWidthZero(_)));

        if let Err(LexerError::NumberLiteralWidthZero(l)) = result {
            assert_eq!(Location::new(1, 1), l);
        }
    }

    #[test]
    fn number_width_too_big() {
        let source_text = "42#70";

        let source = Cursor::new(source_text);

        let mut lexer = Lexer::new(source);

        let result = lexer.get_token();

        assert_matches!(result, Err(LexerError::NumberLiteralWidthTooBig(_)));

        if let Err(LexerError::NumberLiteralWidthTooBig(l)) = result {
            assert_eq!(Location::new(1, 1), l);
        }
    }

    #[test]
    fn number_value_too_big() {
        let source_text = "256#8";

        let source = Cursor::new(source_text);

        let mut lexer = Lexer::new(source);

        let result = lexer.get_token();

        assert_matches!(result, Err(LexerError::NumberLiteralValueTooBig(_)));

        if let Err(LexerError::NumberLiteralValueTooBig(l)) = result {
            assert_eq!(Location::new(1, 1), l);
        }
    }

    #[test]
    #[should_panic(expected = "attempt to add with overflow")]
    fn number_overflow() {
        let source_text = "18446744073709551616";

        let source = Cursor::new(source_text);

        let mut lexer = Lexer::new(source);

        let result = lexer.get_token();

        // TODO: return error before overflow instead of panicking
//        assert_matches!(result, Err(LexerError::NumberLiteralOverflow(_)));
    }

    #[test]
    fn unexpected_character() {
        let source_text = "jane?doe";

        let source = Cursor::new(source_text);

        let mut lexer = Lexer::new(source);

        let result = lexer.get_token();

        assert_matches!(result, Ok(_));

        let result = lexer.get_token();

        assert_matches!(result, Err(LexerError::UnexpectedCharacter(_, _)));

        if let Err(LexerError::UnexpectedCharacter(c, l)) = result {
            assert_eq!('?', c);
            assert_eq!(Location::new(1, 5), l);
        }
    }

    #[test]
    fn all_token_types() {
        let source_text = "\
hello
42
.
,
;
:
=
&
|
^
~
{
}
[
]
(
)
in
out
block
";

        let expected_tokens = vec![
            Token::new(TokenData::Identifier("hello".to_string()), Location::new(1, 1)),
            Token::new(TokenData::Number { value: 42, width: None }, Location::new(2, 1)),
            Token::new(TokenData::Dot, Location::new(3, 1)),
            Token::new(TokenData::Comma, Location::new(4, 1)),
            Token::new(TokenData::Semicolon, Location::new(5, 1)),
            Token::new(TokenData::Colon, Location::new(6, 1)),
            Token::new(TokenData::Equals, Location::new(7, 1)),
            Token::new(TokenData::AND, Location::new(8, 1)),
            Token::new(TokenData::OR, Location::new(9, 1)),
            Token::new(TokenData::XOR, Location::new(10, 1)),
            Token::new(TokenData::NOT, Location::new(11, 1)),
            Token::new(TokenData::LeftBrace, Location::new(12, 1)),
            Token::new(TokenData::RightBrace, Location::new(13, 1)),
            Token::new(TokenData::LeftBracket, Location::new(14, 1)),
            Token::new(TokenData::RightBracket, Location::new(15, 1)),
            Token::new(TokenData::LeftParenthesis, Location::new(16, 1)),
            Token::new(TokenData::RightParenthesis, Location::new(17, 1)),
            Token::new(TokenData::InKeyword, Location::new(18, 1)),
            Token::new(TokenData::OutKeyword, Location::new(19, 1)),
            Token::new(TokenData::BlockKeyword, Location::new(20, 1)),
            Token::new(TokenData::EndOfFile, Location::new(21, 0)),
        ];

        let result = expect_tokens(source_text, &expected_tokens);
        assert!(result.is_ok());
    }

    #[test]
    fn simple_program() {
        let source_text = "\
block and {
    in a[8];
    in b[8];
    out q[8];

    out = a & b;
}";

        let expected_tokens = vec![
            Token::new(TokenData::BlockKeyword, Location::new(1, 1)),
            Token::new(TokenData::Identifier("and".to_string()), Location::new(1, 7)),
            Token::new(TokenData::LeftBrace, Location::new(1, 11)),
            Token::new(TokenData::InKeyword, Location::new(2, 5)),
            Token::new(TokenData::Identifier("a".to_string()), Location::new(2, 8)),
            Token::new(TokenData::LeftBracket, Location::new(2, 9)),
            Token::new(TokenData::Number { value: 8, width: None }, Location::new(2, 10)),
            Token::new(TokenData::RightBracket, Location::new(2, 11)),
            Token::new(TokenData::Semicolon, Location::new(2, 12)),
            Token::new(TokenData::InKeyword, Location::new(3, 5)),
            Token::new(TokenData::Identifier("b".to_string()), Location::new(3, 8)),
            Token::new(TokenData::LeftBracket, Location::new(3, 9)),
            Token::new(TokenData::Number { value: 8, width: None }, Location::new(3, 10)),
            Token::new(TokenData::RightBracket, Location::new(3, 11)),
            Token::new(TokenData::Semicolon, Location::new(3, 12)),
            Token::new(TokenData::OutKeyword, Location::new(4, 5)),
            Token::new(TokenData::Identifier("q".to_string()), Location::new(4, 9)),
            Token::new(TokenData::LeftBracket, Location::new(4, 10)),
            Token::new(TokenData::Number { value: 8, width: None }, Location::new(4, 11)),
            Token::new(TokenData::RightBracket, Location::new(4, 12)),
            Token::new(TokenData::Semicolon, Location::new(4, 13)),
            Token::new(TokenData::OutKeyword, Location::new(6, 5)),
            Token::new(TokenData::Equals, Location::new(6, 9)),
            Token::new(TokenData::Identifier("a".to_string()), Location::new(6, 11)),
            Token::new(TokenData::AND, Location::new(6, 13)),
            Token::new(TokenData::Identifier("b".to_string()), Location::new(6, 15)),
            Token::new(TokenData::Semicolon, Location::new(6, 16)),
            Token::new(TokenData::RightBrace, Location::new(7, 1)),
            Token::new(TokenData::EndOfFile, Location::new(7, 1)),
        ];

        let result = expect_tokens(source_text, &expected_tokens);
        assert!(result.is_ok());
    }
}
